### DISCUSSION TITLE
Use Node.js in actions.js

### DISCUSSION THREAD

**Discussion by JokerMartini** [quasarframework/quasar#9947] [CLI - Electron mode] [answered]
**Updated** question for clarification in hopes to help others who encounter this same issue.

In previous versions of Quasar Electron apps, prior to Quasar 2.0, I used to be able to import the following modules directly into my store's actions.js file, and any additional *.js files that may be used for helper methods. 
```
const fs = require('fs');
const path = require('path');
const { remote } = require('electron');

```
However in Quasar 2.0+ it prevents the use of any Node.js packages from being used which has caused a slew or errors to show. I understand in latest Quasar we can no longer use native Node.js packages for security reasons, however there are good explanations of how to solve this problem. I'm hoping someone can help me solve my problem.

Below are the main bits of various js files I currently use in my Electron Desktop app, that utilize the Node.js packages listed above. Could someone explain to me how I can fix my current files to make them useable again in my project using this newly suggested **electron-preload.js**. Note that I would prefer that I don't have to place all my methods utilizing the Node.js packages into a single js file. That would negate my current organization of code going forward.

**utils.js**
```
const fs = require('fs');
const path = require('path');
const { remote } = require('electron');

async parseConfig (filepath) {
        if (!fs.existsSync(filepath)) {
            return { }
        }
        const obj = JSON.parse(fs.readFileSync(filepath, 'utf-8'))
        ...
}
```
**actions.js**
```
import { shell } from 'electron';
function showFile(filepath) {
      shell.openExternal(filepath);
}
```

Just for clarification these are the errors i now get in Quasar 2.0 when tryin to run dev or build. Which is due to it no long allowing for Node.js packages to be imported.

```
Module not found: Can't resolve imported dependency "fs"
Did you forget to install it? You can run: yarn add fs 

 App •  COMPILATION FAILED  • Please check the log above for details.

Module not found: Can't resolve imported dependency "path"
Did you forget to install it? You can run: yarn add path

 App •  COMPILATION FAILED  • Please check the log above for details.

Module not found: Can't resolve imported dependency "remote"
Did you forget to install it? You can run: yarn add remote

 App •  COMPILATION FAILED  • Please check the log above for details.

Module not found: Can't resolve imported dependency "shell"
Did you forget to install it? You can run: yarn add shell

 App •  COMPILATION FAILED  • Please check the log above for details.
```

**Comment by yusufkandemir** [comment, marked as answer]
Hi, @JokerMartini 👋 Please see https://quasar.dev/start/upgrade-guide#nodejs-polyfills

**Reply by JokerMartini** [reply]
why the --dev flag? will it still allow me to properly do a build of the project?

**Reply by yusufkandemir** [reply]
First of all, you could have just tested it to save us both some time, but anyways. The thing you are installing is a **webpack plugin**, and that's pretty normal for it to be a dev dependency. It does some magic so you can still import Nodejs core modules. As long as you are not importing anything from the `node-polyfill-webpack-plugin`, you can properly build the project.

**Reply by JokerMartini** [reply]
I understand, i appreciate the help. I hope over the coming weeks that more information on how to use preload come to light. The examples are pretty limited. I don't understand how they are suppose to be used in projects that previously contain multiple custom js files that import node.js packages like mentioned above.

**Reply by yusufkandemir** [reply]
Considering the lack of context, I can't say that I understood much, but you can check Webpack's docs https://webpack.js.org/guides/code-splitting/#prefetchingpreloading-modules

**Reply by JokerMartini** [reply]
I can explain. Here is my folder structure... 
My settings.js and utils.js files both need to be called from the store/actions.js. However the settings.js and utils.js files both need to import modules like the the ones listed here. However I get errors...

```
const fs = require('fs');
const path = require('path');
const { remote } = require('electron');
import os from 'os';
import { resolve } from 'dns';
const { ipcRenderer } = require('electron');
```

**Errors**
```
App •  DONE  • "Renderer" compiled with errors • 20225ms
App •  ERROR  •  Renderer  in ./src/helpers/utils.js

Module not found: Can't resolve imported dependency "fs"       
Did you forget to install it? You can run: yarn add fs

App •  ERROR  •  Renderer  in ./src/helpers/utils.js

Module not found: Can't resolve imported dependency "dns"      
Did you forget to install it? You can run: yarn add dns        

App •  ERROR  •  Renderer  in ./node_modules/electron/index.js

Module not found: Can't resolve imported dependency "fs"
Did you forget to install it? You can run: yarn add fs

App •  COMPILATION FAILED  • Please check the log above for details.
```

![image](https://user-images.githubusercontent.com/1416749/124392632-be56ea80-dcc4-11eb-84cd-fb330eedb893.png)

**Reply by JokerMartini** [reply]
If you have a time to spare. id be more than glad to share the project with ya and pay you for your time, if you could just help me get initially setup and working properly. I can then do the rest, just need help getting started. It's in BitBucket.

**Reply by yusufkandemir** [reply]
Ah, so you meant preload as in Electron. Please check this guide, and let me know if it went well: https://quasar.dev/quasar-cli/developing-electron-apps/electron-upgrade-guide

**Reply by JokerMartini** [reply]
Yep luckily I created the project just the other day so it's setup with all this already by the default cli commands. However I'm not clear on how to write my code to work in the new structure where security improvements have changed. I still need help on that.

**Reply by JokerMartini** [reply]
Your willingness to help is appreciated. And my folder structure you see above is where I'm stuck.

**Reply by JokerMartini** [reply]
@yusufkandemir  I've gone ahead and updated the original question to clarify what the problem is. Hope you can help. Thank you.

**Reply by yusufkandemir** [reply]
You can still have multiple files, as long as you import and use them from the Electron preload script. You can take a look at the docs for more info: https://quasar.dev/quasar-cli/developing-electron-apps/electron-preload-script
It suggests using `contextBridge`, you can get more information about it on [Electron Docs](https://www.electronjs.org/docs/api/context-bridge).
I don't see how you import and use the functions you've shown in your UI, but you can do something like this:

```js
// In Preload script
import { showFile } from 'src/actions' // or wherever the actions.js file is
// ...
contextBridge.exposeInMainWorld('actions', {
  showFile,
})

// In UI Code
window.actions.showFile('path')
```

**Reply by JokerMartini** [reply]
The code i wrote above was just pseudo code. Sorry if it didn't read correctly, but i believe what you have shown here showcases exactly what i needed. I'm going to give it a shot and check it out right now. Thank you very much. ill report back after i test it :)

**Reply by JokerMartini** [reply]
@yusufkandemir  I tried what you said but still get an import error. I'm not not why it can't resolve this path?
![image](https://user-images.githubusercontent.com/1416749/124677093-bac28f80-de8d-11eb-947d-088dba42ad1c.png)

**Reply by yusufkandemir** [reply]
@JokerMartini can you try using a relative path instead `../src/utils`, so we can be sure that importing from outside works. Please let me know if that one work, after that I will take a look at Webpack aliases to also make `src/utils` work.

**Reply by JokerMartini** [reply]
works like a charm. thank you for all the help. I would suggesting putting something along these lines in the help docs so people better understand a real world example.

Thank you!!
